### ISSUE TITLE
Consider renaming the API ThrowOnUnexpectedArgument

### ISSUE THREAD

**Issue by parekhkb** [natemcmaster/CommandLineUtils#88]
I am trying to make my app not throw an exception if the user enters an invalid command/option/argument. 

I set ThrowOnUnexpectedArgument to false in the constructor of the CommandLineApplication. I also set a ValidationErrorHandler function.

However, an exception is still thrown.

![image](https://user-images.githubusercontent.com/5116394/39063670-aa388110-4480-11e8-9316-9c74604aef6d.png)

**Comment by natemcmaster** [comment]
Can you provide repro steps? I have a guess, but it's hard to tell without knowing what is being passed in as `args[]` and what  CommandBase.ConfigurationApplication is doing in your code.

**Comment by parekhkb** [comment]
Im calling `dotnet MyApp.dll foo` or `dotnet MyApp.dll --foo` both repro the issue.

Below are some relevant code snippets:


```C#
        private void ConfigureCommand(CommandLineApplication command, IServiceCollection services)
        {
            _services = services;

            command.Description = Description;
            command.HelpOption("-?|-h|--help");

            ConfigureCommandInternal(command);

            command.OnExecute(() => OnExecute(command));
        }
```

```C#
protected override void ConfigureCommandInternal(CommandLineApplication command)
        {
            // Register sub-commands
            RegisterSubCommand<MigrateMasterCommand>(command);
            RegisterSubCommand<MigratePublicCommand>(command);
            RegisterSubCommand<MigratePrivateCommand>(command);
            RegisterSubCommand<MigrateAllCommand>(command);
        }
```

```C#
        public CommandLineApplication RegisterSubCommand<TCommand>(TCommand com, CommandLineApplication root) where TCommand : CommandBase
        {
            return root.Command(com.Name.ToLower(), c => com.ConfigureCommand(c, _services));
        }

        public CommandLineApplication RegisterSubCommand<TCommand>(CommandLineApplication root) where TCommand : CommandBase, new()
        {
            return RegisterSubCommand(new TCommand(), root);
        }
```

**Comment by natemcmaster** [comment]
Try changing this:
```diff
        public CommandLineApplication RegisterSubCommand<TCommand>(TCommand com, CommandLineApplication root) where TCommand : CommandBase
        {
-            return root.Command(com.Name.ToLower(), c => com.ConfigureCommand(c, _services));
+            return root.Command(com.Name.ToLower(), 
+              c => com.ConfigureCommand(c, _services),
+              throwOnUnexpectedArg: true);
        }
```

**Comment by natemcmaster** [comment]
Typo: I meant `throwOnUnexpectedArg: false`

**Comment by parekhkb** [comment]
Thanks, @natemcmaster, that was indeed the issue. However I think there might be another bug: If I use throwOnUnexpectedArg: false. Then My OnValidationError Handler never get's called, and the app returns 0 rather than non-zero. I ended up just catching the ComandParsingException myself in order to  write a useful message, eat the error, and return -1.

**Comment by natemcmaster** [comment]
What kind of validators have you configured? The app should return `1` by default if there are validation errors. ValidationErrorHandler is not invoked if there were not validation errors.

By the way, each subcommand is a new instance of CommandLineApplication, each with its own ValidationErrorHandler. Are you invoking a subcommand, but expecting the validation error handler on the root CommandLineApplication object to be called?

**Comment by parekhkb** [comment]
I see. Thanks for your help. I guess the issue I have with this feature is that it doesn't exit execution onValidationError.

**Comment by natemcmaster** [comment]
I think there is probably a bug here somewhere, I'm just not sure what it is. So far, I've identified at least two things that are not clear about the behavior of this API

 - throwOnUnexpectedArgs has to applied to all subcommands explicitly. A better default may be to honor the value set on the root command instead of defaulting to false
 - ValidationErrorHandler is only invoked on the command with the validation error, so you end up re-defining a customer error handler on all subcommand objects

Does this describe your issues so far? I'm not sure what you meant by the most recent comment about "it doesn't exit execution onValidationError."

**Comment by parekhkb** [comment]
Here's what I mean by it doesn't exit exuection onValidationError. I set throwOnUnexpectedArgs to false for the root and all subcommands, then I specify an OnValidation() handler for each command, my handler returns 3 (non-zero). I would expect the app to exit after the onValidation handler is executed, but it instead proceeds to execute my app. For example. If I call `dotnet MyApp.dll ValidCommand --valid-option --invalid-option`, i would expect the existence of invalid option to prevent the app from doing what it would do if I just ran `dotnet MyApp.dll ValidCommand --valid-option`

**Comment by natemcmaster** [comment]
Oh, I see. Ok, so maybe this explanation will help.

What ThrowOnUnexpectedArgs:false means to the parser is "parse until you hit an unrecognized argument or option. Then, but everything else into the CommandLineApplication.RemainingArguments property".

For example:

Input | ValidOption.HasValue() | RemainingArguments (string[])
--|--|--
`ValidCommand --valid-option --invalid-option` | `true` | `{ "--invalid-option" }`
`ValidCommand --invalid-option --valid-option` | `false` | `{ "--invalid-option", "--valid-option" }`
`ValidCommand banana --invalid-option --valid-option` | `false` | `{ "banana", "--invalid-option", "--valid-option" }`

This behavior was added before this library every had support for validation on arguments. I realize now why this behavior isn't obvious. It's probably time to reconsider what the "throwOnUnexpectedArgs" value means.

**Comment by natemcmaster** [comment]
Stale bot closed this, but I'd like to revisit this in 3.0. Reopening

**Comment by natemcmaster** [comment]
Finally got around to this. I've renaming the API. See https://github.com/natemcmaster/CommandLineUtils/pull/340 and #339. My plan is to obsolete the API in 2.6 and delete in 3.0. I'll be merging the changes soon to close this out, but it will probably take a few weeks to make any releases to nuget.org.
